File: matlab/road_grading_data.m
% Least-cost road grading.

n = 100;
e = 5*sin((1:n)/n*3*pi)'+sin((1:n)/n*10*pi)';% elevation of the road
d = 1; % distance between points
D1 = .08; % the road grade should never be greater than 8%
D2 = .025; % the road grade should never change faster than 25% over 10 meters
D3 = .005; % a further constraint on the smoothness of the road.

% cut and fill function coefficients of the form alpha*x_+^2+beta*x_+
alpha_fill = 2;
beta_fill = 30;
alpha_cut = 12;
beta_cut = 1;

% a plot of the cost functions
figure
elevation = 0:.1:10;
plot(elevation,alpha_fill*elevation.^2+beta_fill*elevation,'b');
hold on
plot(elevation,alpha_cut*elevation.^2+beta_cut*elevation,'g');
xlabel('elevation change')
ylabel('cost')
legend('fill','cut')

print -depsc road_grading_cost_function.eps

h = ones(n,1)*mean(e); % replace h with your solution.

figure
subplot(2,1,1)
plot((0:n-1)*d,e,'--r');
ylabel('elevation');
hold on
plot((0:n-1)*d,h, 'b')
legend('e','h');
subplot(2,1,2)
plot((0:n-1)*d,h-e)
ylabel('elevation change')
xlabel('distance')

File: python/road_grading_data.py
import numpy as np
import matplotlib.pyplot as plt

# Least-cost road grading

n = 100
d = 1  # distance between points

# Elevation of the road
e = 5 * np.sin(np.arange(1, n + 1) / n * 3 * np.pi) + np.sin(
    np.arange(1, n + 1) / n * 10 * np.pi
)

# Constraints
D1 = 0.08  # the road grade should never be greater than 8%
D2 = 0.025  # the road grade should never change faster than 25% over 10 meters
D3 = 0.005  # a further constraint on the smoothness of the road.

# Cut and fill function coefficients
alpha_fill = 2
beta_fill = 30
alpha_cut = 12
beta_cut = 1

# Plot cost functions
elevation = np.arange(0, 10.1, 0.1)
cost_fill = alpha_fill * elevation**2 + beta_fill * elevation
cost_cut = alpha_cut * elevation**2 + beta_cut * elevation

plt.figure()
plt.plot(elevation, cost_fill, "b", label="fill")
plt.plot(elevation, cost_cut, "g", label="cut")
plt.xlabel("Elevation Change")
plt.ylabel("Cost")
plt.legend()
plt.grid()
plt.savefig("road_grading_cost_function.eps")
plt.show()

if __name__ == "__main__":
    # Replace h with your solution
    h = np.ones(n) * np.mean(e)

    # Plot elevation profiles
    plt.figure(figsize=(10, 6))

    plt.subplot(2, 1, 1)
    plt.plot(np.arange(n) * d, e, "--r", label="e")
    plt.plot(np.arange(n) * d, h, "b", label="h")
    plt.ylabel("Elevation")
    plt.legend()
    plt.grid()

    plt.subplot(2, 1, 2)
    plt.plot(np.arange(n) * d, h - e)
    plt.ylabel("Elevation Change")
    plt.xlabel("Distance")
    plt.grid()

    plt.show()
